File: src/loop.ts
import { randomUUID } from 'node:crypto';
import { relative, resolve } from 'node:path';
import type {
  RunConfig, RunState, Finding, Patch, Attempt, LoopEvent, Phase, CheckResult,
} from './types.js';
import { Ledger, ledgerPathFor } from './ledger.js';
import { proposePatches } from './propose.js';
import { runCheck } from './checks.js';
import { verifyPatch } from './verify.js';
import { buildImportGraph, scopeOf } from './imports.js';
import { byRisk, suppressFindings } from './risk.js';
import { applyEdits } from './patch.js';
import { createProvider, CRITIC_QUESTIONS, type Provider } from './provider.js';
import { Tracer } from './tracer.js';
import { execute, type WorkNode } from './graph.js';
import { inspect, useBranch, commitFile, logSince, head } from './git.js';

type Emit = (e: LoopEvent) => void;

/** Which loop phase each graph node reports as. */
const PHASE_OF = (id: string): Phase =>
  id.startsWith('observe') ? 'observe'
  : id === 'reduce' || id === 'diagnose' ? 'diagnose'
  : id === 'propose' ? 'repair'
  : 'verify';

/**
 * The loop, executed as a graph of agents.
 *
 * Each iteration deploys a small fleet: one observer agent per check (all
 * concurrent), a diagnoser, a proposer, three critic agents with fresh
 * contexts (concurrent), a majority gate, and a single serial verify tail.
 *
 *   observe(test)  observe(typecheck)  observe(lint)   ← independent, concurrent
 *            └─────────────┬──────────────┘
 *                       reduce                          ← plain code, no model
 *                          │
 *                       diagnose                        ← rank, consult the ledger
 *                          │
 *                       propose                         ← the only creative step
 *                          │
 *            ┌─────────────┼─────────────┐
 *         correct?    collateral?      valid?           ← fresh context each
 *            └─────────────┼─────────────┘
 *                        gate                           ← majority vote
 *                          │
 *                       verify                          ← apply, re-run checks, decide
 *
 * The tail is serial and must stay that way: two patches applied at once
 * destroy the verify step's ability to say which one caused what. That edge
 * is a real dependency, not a leftover from writing the steps in order.
 */
export class Loop {
  private state: RunState;
  private ledger: Ledger;
  private provider: Provider | null = null;
  private tracer: Tracer = new Tracer();
  /** HEAD before the run, so the final report can list only our commits. */
  private gitBase: string | null = null;
  /** Findings with no untried candidates, quarantined for THIS run only. */
  private quarantinedThisRun = new Set<string>();

  constructor(private config: RunConfig, private emit: Emit) {
    this.state = {
      id: randomUUID().slice(0, 8),
      config,
      findings: [],
      attempts: [],
      iteration: 0,
      status: 'idle',
      startedAt: new Date().toISOString(),
    };
    this.ledger = new Ledger(config.statePath ?? ledgerPathFor(config.sourceRoot));
  }

  get snapshot(): RunState {
    return JSON.parse(JSON.stringify(this.state));
  }

  /** Findings still worth working on, given what the ledger has learned. */
  actionableRemaining(): Finding[] {
    return this.state.findings.filter((f) => !this.ledger.isExhausted(f));
  }

  private say(phase: Phase, message: string, data?: unknown) {
    this.emit({
      runId: this.state.id,
      iteration: this.state.iteration,
      phase,
      at: new Date().toISOString(),
      message,
      data,
    });
  }

  async run(): Promise<RunState> {
    this.state.status = 'running';
    this.tracer = await Tracer.create();
    this.tracer.startRun({
      sourceRoot: this.config.sourceRoot,
      checks: this.config.checks.map((c) => c.name),
      budget: this.config.budget,
    });
    try {
      // Before touching anything: if we are going to commit, the tree has to
      // be clean first, or the person cannot tell our edits from their own.
      if (this.config.git) {
        const repo = await inspect(this.config.sourceRoot);
        if (!repo.isRepo) {
          throw new Error(`--git was set but ${this.config.sourceRoot} is not a git repository`);
        }
        if (!repo.clean) {
          throw new Error(
            `--git needs a clean tree; ${repo.dirty.length} path(s) have uncommitted changes ` +
            `(${repo.dirty.slice(0, 3).join(', ')}${repo.dirty.length > 3 ? ', …' : ''}). ` +
            `Commit or stash them first — otherwise Kintsugi's edits and yours end up in the same diff.`,
          );
        }
        this.gitBase = await head(this.config.sourceRoot);
        const branch = this.config.branch ?? 'kintsugi/fixes';
        await useBranch(this.config.sourceRoot, branch);
        this.say('settle', `Committing to branch ${branch} (was on ${repo.branch ?? 'detached HEAD'})`);
      }

      // Establish up front whether the model path works, rather than letting
      // a bad credential masquerade as the model having nothing to suggest.
      const provider = await createProvider(this.config);
      if (!provider) {
        this.say('settle', 'No model configured — running rules-only (every patch is still verified the same way)');
      } else if (provider.name === 'mock') {
        this.provider = provider;
        this.say('settle', 'Model proposer: mock — rules first, replayed proposals for what rules cannot reach');
      } else {
        const check = await (provider as any).preflight();
        if (check.ok) {
          this.provider = provider;
          this.say('settle', `Model proposer ${check.detail} — rules first, model for what rules cannot reach`);
        } else {
          this.say('settle', `Model proposer unavailable: ${check.detail}`);
          this.say('settle', 'Continuing rules-only. Findings with no mechanical rule will be reported, not fixed.');
        }
      }

      for (let i = 1; i <= this.config.maxIterations; i++) {
        this.state.iteration = i;
        const settled = await this.iterate();
        if (settled) break;
      }

      if (this.state.status === 'running') {
        this.state.status = 'exhausted';
        this.say('settle', `Iteration budget (${this.config.maxIterations}) reached`);
      }
    } catch (err) {
      this.state.status = 'failed';
      this.say('settle', `Run failed: ${(err as Error).message}`);
    } finally {
      // A final settle span mirroring the ledger's attempt records — one
      // entry per attempt with the same outcome vocabulary — so the trace
      // and the ledger tell the same story and are joinable on
      // fingerprint + outcome.
      try {
        this.tracer.span('settle', {
          status: this.state.status,
          iterations: this.state.iteration,
          attempts: this.state.attempts.map((a) => ({
            fingerprint: a.fingerprint,
            outcome: a.outcome,
            patch: { file: a.patch?.file, rationale: a.patch?.rationale },
            provider: a.provider,
            collateral: a.collateral,
            at: a.at,
          })),
        });
      } catch {
        // best effort
      }
      this.tracer.flush();
      if (this.config.git && this.gitBase) {
        const commits = await logSince(this.config.sourceRoot, this.gitBase);
        this.say('settle', commits.length
          ? `${commits.length} commit(s) on this branch — review with: git log -p ${this.gitBase.slice(0, 7)}..HEAD`
          : 'No commits — nothing was verified as safe to apply');
        for (const c of commits) this.say('settle', `  ${c}`);
      }
      this.state.endedAt = new Date().toISOString();
    }

    return this.state;
  }

  /** Build and run one iteration's graph. Returns true when the loop should stop. */
  private async iterate(): Promise<boolean> {
    const { checks, sourceRoot, dryRun } = this.config;
    let settled = false;

    const nodes: WorkNode[] = [];

    // ---- fan out: one observer agent per check --------------------------
    for (const check of checks) {
      nodes.push({
        id: `observe:${check.name}`,
        job: `Run ${check.name} (${check.command})`,
        dependsOn: [],
        validate: (o: any) => !!o && typeof o === 'object' && 'findings' in o,
        run: async ({ emit }) => {
          const result = await runCheck(check, sourceRoot);
          this.tracer.span('observe', {
            check: check.name,
            durationMs: result.durationMs,
            findings: result.findings.length,
            crashed: result.crashed,
          });
          if (result.crashed) emit(`crashed (exit ${result.exitCode}) — a broken harness, not a defect; not healing`);
          else if (result.findings.length === 0) emit(`clean (${result.durationMs}ms)`);
          else emit(`${result.findings.length} finding(s) (${result.durationMs}ms)`);
          return result;
        },
      });
    }

    // ---- converge: plain code, no model ---------------------------------
    nodes.push({
      id: 'reduce',
      job: 'Merge check results into one deduplicated finding list',
      dependsOn: checks.map((c) => `observe:${c.name}`),
      validate: (o: any) => !!o && Array.isArray(o.findings) && o.baseline instanceof Set,
      run: async ({ deps, emit }) => {
        const findings: Finding[] = [];
        const seen = new Set<string>();
        for (const c of checks) {
          const result = deps[`observe:${c.name}`] as CheckResult;
          for (const f of result.findings) {
            if (seen.has(f.fingerprint)) continue;
            seen.add(f.fingerprint);
            findings.push(f);
          }
        }
        // Suppression (generated code, test-file style) filters the queue
        // before ranking; risk scoring then orders worst-first within each
        // severity band. Both engines apply the same arithmetic, so the
        // order they converge in is identical.
        const { kept, dropped } = suppressFindings(findings);
        if (dropped.length) {
          emit(`${dropped.length} finding(s) suppressed (generated, test-file style, or expected domain complexity)`);
        }
        kept.sort(byRisk);
        this.state.findings = kept;
        return { findings: kept, baseline: new Set(kept.map((f) => f.fingerprint)) };
      },
    });

    // ---- diagnose --------------------------------------------------------
    nodes.push({
      id: 'diagnose',
      job: 'Rank findings worst-first and pick one target',
      dependsOn: ['reduce'],
      validate: (o: any) => !!o && 'target' in o,
      run: async ({ deps, emit }) => {
        const { findings } = deps.reduce as { findings: Finding[] };
        const actionable = findings
          .filter(
            (f) =>
              !this.ledger.isExhausted(f) &&
              // Within one run, a finding with no untried candidates is
              // quarantined for this run — but only the ledger decides
              // whether that is permanent. Without this, a rules-only run
              // would re-target the same dead end every iteration.
              !this.quarantinedThisRun.has(f.fingerprint),
          )
          .sort(byRisk);

        const quarantined = findings.length - actionable.length;
        if (quarantined > 0) {
          emit(`${quarantined} finding(s) already quarantined — left for a human`);
        }
        if (actionable.length === 0) {
          emit('Nothing actionable left. Converged.');
          return { target: null };
        }
        emit(`Targeting ${actionable[0].check}: ${actionable[0].summary}`, actionable[0]);
        return { target: actionable[0] };
      },
    });

    // ---- propose ---------------------------------------------------------
    nodes.push({
      id: 'propose',
      job: 'Produce candidate patches for the target finding',
      dependsOn: ['diagnose'],
      validate: (o: any) => Array.isArray(o?.candidates),
      run: async ({ deps, emit }) => {
        const target = (deps.diagnose as any).target as Finding | null;
        if (!target) return { candidates: [], target: null };

        // Rules first: free, instant, and proven on these classes.
        let candidates = await proposePatches(target, sourceRoot);

        // The model is consulted only for what the rules could not reach.
        if (candidates.length === 0 && this.provider) {
          emit(`No rule covers ${target.code ?? target.check} — asking ${this.provider.name}`);
          try {
            candidates = await this.provider.propose(target, sourceRoot);
            emit(`${this.provider.name} proposed ${candidates.length} candidate(s)`);
            this.tracer.generation(
              'propose',
              (this.provider as any)?.lastUsage ?? null,
              { fingerprint: target.fingerprint, check: target.check, candidates: candidates.length },
            );
          } catch (err) {
            emit(`Model proposer failed: ${(err as Error).message}`);
          }
        }

        // Blast radius is decided now, from what the file is — a repair that
        // touches a module other files import changes code the loop is not
        // looking at, and the verify gate cannot catch that.
        const graph = buildImportGraph(sourceRoot);
        for (const p of candidates) {
          const { scope, importers } = scopeOf(graph, p.file);
          p.scope = scope;
          p.blastRadius = importers;
        }

        const viable = this.ledger.prioritise(target.fingerprint, candidates);
        return { candidates: viable, target };
      },
    });

    // ---- three critic agents, fresh context, in parallel -----------------
    CRITIC_QUESTIONS.forEach((question, idx) => {
      nodes.push({
        id: `critic:${idx}`,
        job: question,
        dependsOn: ['propose'],
        validate: (o: any) => !!o && (o.verdict === 'keep' || o.verdict === 'drop'),
        run: async ({ deps, emit }) => {
          const { candidates, target } = deps.propose as { candidates: Patch[]; target: Finding | null };
          // Nothing to judge, or no judge available. Abstaining as "keep" is
          // safe because the deterministic verify gate still runs afterwards.
          if (!candidates.length || !target || !this.provider?.critique) {
            return { verdict: 'keep', reason: 'not consulted' };
          }
          const out = await this.provider.critique(candidates[0], target, question);
          if (!out) return { verdict: 'keep', reason: 'no verdict returned' };
          if (out.verdict === 'drop') emit(`drop — ${out.reason}`);
          return out;
        },
      });
    });

    // ---- gate ------------------------------------------------------------
    nodes.push({
      id: 'gate',
      job: 'Majority vote across the three checks',
      dependsOn: CRITIC_QUESTIONS.map((_, i) => `critic:${i}`),
      validate: (o: any) => typeof o?.passed === 'boolean',
      run: async ({ deps, emit }) => {
        const verdicts = CRITIC_QUESTIONS.map((_, i) => (deps[`critic:${i}`] as any).verdict);
        const drops = verdicts.filter((v) => v === 'drop').length;
        const passed = drops < Math.ceil(verdicts.length / 2);
        if (drops > 0) emit(`${drops}/${verdicts.length} checks voted drop — ${passed ? 'proceeding' : 'rejecting'}`);
        return { passed };
      },
    });

    // ---- verify: the serial tail -----------------------------------------
    nodes.push({
      id: 'verify',
      job: 'Apply, re-run the checks, and either commit or revert',
      dependsOn: ['reduce', 'propose', 'gate'],
      validate: (o: any) => typeof o?.settled === 'boolean',
      run: async ({ deps, emit }) => {
        const { candidates, target } = deps.propose as { candidates: Patch[]; target: Finding | null };
        const { passed } = deps.gate as { passed: boolean };
        const { baseline } = deps.reduce as { baseline: Set<string> };

        if (!target) return { settled: true, converged: true };

        if (!candidates.length) {
          emit('No untried patch available for this finding — quarantining for a human');
          this.quarantinedThisRun.add(target.fingerprint);
          this.record(target, null, 'unverifiable', []);
          return { settled: false };
        }
        if (!passed) {
          emit('Critic agents rejected the patch — recording and moving on');
          this.record(target, candidates[0], 'unverifiable', []);
          return { settled: false };
        }
        if (dryRun) return this.drySurvey(emit);

        const patch = candidates[0];

        // A shared file is out of scope for an automatic fix, however
        // cleanly it would clear the check. Editing it moves code the loop
        // is not looking at, which is a decision its owner makes — and the
        // verify gate cannot catch this, because the finding really does
        // clear. Escalating is the correct outcome, not a limitation.
        if (patch.scope === 'shared' && !this.config.allowShared) {
          emit(`ESCALATED — ${patch.rationale}`);
          emit(`Not applied: ${relative(sourceRoot, patch.file)} is imported by ${patch.blastRadius ?? '?'} ` +
            `module(s). Editing it reaches past the defect. Re-run with --allow-shared to override.`);
          this.record(target, patch, 'unverifiable', []);
          return { settled: false };
        }
        if (patch.scope === 'shared') {
          emit(`Shared-file change permitted by --allow-shared (${patch.blastRadius} importer(s))`);
        }

        emit(`Patch: ${patch.rationale}`);

        const restore = applyEdits([patch, ...(patch.also ?? [])], sourceRoot);
        let outcome: Attempt['outcome'];
        let collateral: string[] = [];
        let collateralDetail: string[] = [];

        try {
          const v = await verifyPatch(checks, sourceRoot, baseline, target.fingerprint);
          outcome = v.outcome;
          collateral = v.collateral.map((f) => f.fingerprint);
          collateralDetail = v.collateral.map((f) => `${f.check}: ${f.summary}`);
          // The span mirrors the ledger's attempt record: same fingerprint,
          // same patch identity, same outcome vocabulary — so a trace is
          // queryable by finding the way the ledger is.
          this.tracer.span('verify', {
            fingerprint: target.fingerprint,
            check: target.check,
            code: target.code ?? '',
            patch: {
              file: patch.file,
              find: patch.find,
              replace: patch.replace,
              rationale: patch.rationale,
            },
            outcome: v.outcome,
            collateral: collateralDetail,
            provider: !!this.provider,
            durationMs: v.runs.reduce((s, r) => s + r.durationMs, 0),
          });
        } catch (err) {
          // A failed verification proves nothing, so it cannot count as
          // success. Treat it as a failure and put the file back.
          emit(`Verification failed: ${(err as Error).message}`);
          outcome = 'unverifiable';
        }

        if (outcome !== 'committed') {
          restore();
          emit(`${outcome} — reverted`, { collateral });
          for (const c of collateralDetail) emit(`  caused: ${c}`);
        } else {
          emit('committed — finding cleared with no collateral');

          // One commit per verified fix, so each can be read, kept, or
          // dropped on its own merits rather than as an all-or-nothing blob.
          if (this.config.git) {
            try {
              const sha = await commitFile(
                this.config.sourceRoot,
                patch.file,
                `Fix ${target.check}: ${target.summary.slice(0, 60)}`,
                `${target.summary}\n\n${patch.rationale}\n\nVerified by re-running the checks after ` +
                `the change: the finding cleared and no new finding appeared.`,
              );
              if (sha) emit(`  ${sha} committed to git`);
            } catch (err) {
              emit(`  git commit failed (the fix is still applied): ${(err as Error).message}`);
            }
          }
        }

        this.record(target, patch, outcome, collateral);
        return { settled: false };
      },
    });

    // ---- run the graph ---------------------------------------------------
    const report = await execute(nodes, {
      emit: (nodeId, message, data) => this.say(PHASE_OF(nodeId), message, data),
    });

    const layerShape = report.layers.map((l) => l.length).join('→');
    this.say('observe', `Graph: ${report.layers.length} layers (${layerShape}) — ` +
      `${checks.length} observer agent(s) ran concurrently`);

    for (const r of Object.values(report.results)) {
      if (r.status === 'invalid') {
        this.say('repair', `Node ${r.id} returned a result that did not match its contract after ${r.attempts} attempt(s) — discarded`);
      } else if (r.status === 'failed') {
        this.say('repair', `Node ${r.id} failed: ${r.error}`);
      }
    }

    const verify = report.results.verify;
    if (verify?.status === 'ok') {
      const out = verify.output as { settled: boolean; converged?: boolean };
      if (out.converged) {
        this.state.status = 'converged';
        this.say('settle', 'Converged.');
        settled = true;
      }
    } else if (verify?.status === 'skipped') {
      // The tail could not run because something upstream failed. Stopping is
      // correct: continuing would loop on a graph that cannot reach a verdict.
      this.state.status = 'failed';
      this.say('settle', `Iteration could not complete: ${verify.error}`);
      settled = true;
    }

    return settled;
  }

  /** Dry-run survey: report every finding's mechanical answer, write nothing. */
  private async drySurvey(
    emit: (message: string, data?: unknown) => void,
  ): Promise<{ settled: boolean; converged: boolean }> {
    let withPatch = 0;
    let escalated = 0;
    let none = 0;
    for (const f of this.state.findings) {
      const cands = await proposePatches(f, this.config.sourceRoot);
      if (!cands.length) {
        none++;
        emit(`NO PATCH   ${f.check}: ${f.summary}`);
        continue;
      }
      const shared = cands[0].scope === 'shared' && !this.config.allowShared;
      if (shared) {
        escalated++;
        emit(`ESCALATE   ${f.check}: ${f.summary}`);
        emit(`           → ${relative(this.config.sourceRoot, cands[0].file)} — shared file, not applied automatically`);
      } else {
        withPatch++;
        emit(`WOULD PATCH ${f.check}: ${f.summary}`);
        emit(`           → ${relative(this.config.sourceRoot, cands[0].file)} — ${cands[0].rationale}`);
      }
    }
    emit(`Dry run: ${withPatch} patchable, ${escalated} escalated, ${none} with no mechanical answer. Nothing written.`);
    return { settled: true, converged: true };
  }

  private record(
    finding: Finding,
    patch: Patch | null,
    outcome: Attempt['outcome'],
    collateral: string[],
  ) {
    const attempt: Attempt = {
      fingerprint: finding.fingerprint,
      patch: patch ?? {
        id: 'none', file: '', find: '', replace: '',
        rationale: `no candidate patch for ${finding.check}: ${finding.summary}`,
        scope: 'local',
      },
      outcome,
      at: new Date().toISOString(),
      collateral,
      provider: !!this.provider,
    };
    this.ledger.record(attempt);
    this.state.attempts.push(attempt);
  }
}

File: py/kintsugi/loop.py
"""The repair loop — the Python engine's deterministic core.

The TypeScript engine deploys an agent graph (concurrent observers, critic
agents, a majority gate) on top of the same semantics; the Python engine is
the sequential core: observe (run every check), diagnose (rank worst-first,
consult the ledger), propose (mechanical rules), verify (apply, re-run the
checks, keep or revert), settle (converge or exhaust).

The tail is serial and must stay that way: two patches applied at once
destroy the verify step's ability to say which one caused what.
"""

import os
import time
import uuid
from concurrent.futures import ThreadPoolExecutor
from functools import cmp_to_key

from .checks import run_check
from .git import commit_file, head, inspect, use_branch
from .imports import build_import_graph, scope_of
from .ledger import Ledger, ledger_path_for
from .patch import apply_edits
from .propose import propose_patches
from .provider import CRITIC_QUESTIONS, create_provider
from .risk import by_risk, suppress_findings
from .tracer import Tracer
from .verify import verify_patch


def _iso_now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


class Loop:
    def __init__(self, config: dict, emit):
        self.config = config
        self.emit = emit
        self.state = {
            "id": uuid.uuid4().hex[:8],
            "config": config,
            "findings": [],
            "attempts": [],
            "iteration": 0,
            "status": "idle",
            "startedAt": _iso_now(),
        }
        self.ledger = Ledger(config.get("statePath") or ledger_path_for(config["sourceRoot"]))
        self.git_base = None
        self.quarantined_this_run = set()
        self.provider = None
        self.tracer = Tracer.create()

    def say(self, phase: str, message: str) -> None:
        self.emit({
            "runId": self.state["id"],
            "iteration": self.state["iteration"],
            "phase": phase,
            "at": _iso_now(),
            "message": message,
        })

    def actionable_remaining(self):
        return [f for f in self.state["findings"] if not self.ledger.is_exhausted(f)]

    # ------------------------------------------------------------- run

    def run(self):
        self.state["status"] = "running"
        source_root = self.config["sourceRoot"]

        # Before touching anything: if we are going to commit, the tree has
        # to be clean first, or the person cannot tell our edits from theirs.
        if self.config.get("git"):
            repo = inspect(source_root)
            if not repo["isRepo"]:
                raise RuntimeError(f"--git was set but {source_root} is not a git repository")
            if not repo["clean"]:
                dirty = repo["dirty"][:3]
                more = ", …" if len(repo["dirty"]) > 3 else ""
                raise RuntimeError(
                    f"--git needs a clean tree; {len(repo['dirty'])} path(s) have uncommitted "
                    f"changes ({', '.join(dirty)}{more}). Commit or stash them first — otherwise "
                    "Kintsugi's edits and yours end up in the same diff."
                )
            self.git_base = head(source_root)
            branch = self.config.get("branch") or "kintsugi/fixes"
            use_branch(source_root, branch)
            self.say("settle", f"Committing to branch {branch} (was on {repo.get('branch') or 'detached HEAD'})")

        # Establish up front whether the model path works, rather than
        # letting a bad credential masquerade as the model having nothing to
        # suggest. A config may inject a provider object directly (tests);
        # otherwise build one from --llm-mock or the installed SDK.
        self.provider = self.config.get("provider") or create_provider(self.config)
        if self.provider is None:
            self.say("settle", "No model configured — running rules-only (every patch is still verified the same way)")
        elif self.provider.name == "mock":
            self.say("settle", "Model proposer: mock — rules first, replayed proposals for what rules cannot reach")
        else:
            check = self.provider.preflight()
            if check.get("ok"):
                self.say("settle", f"Model proposer {check['detail']} — rules first, model for what rules cannot reach")
            else:
                self.say("settle", f"Model proposer unavailable: {check['detail']}")
                self.say("settle", "Continuing rules-only. Findings with no mechanical rule will be reported, not fixed.")
                self.provider = None

        self.tracer.start_run(self.config)

        checks = self.config["checks"]
        runs = self._run_all(checks, source_root)
        for r in runs:
            self.tracer.span(
                "observe",
                check=r["check"],
                durationMs=r["durationMs"],
                findings=len(r["findings"]),
                crashed=r["crashed"],
            )
        findings = [f for r in runs for f in r["findings"]]
        # Suppression (generated code, test-file style) filters the queue
        # before ranking; risk scoring then orders worst-first within each
        # severity band. Both engines apply the same arithmetic, so the
        # order they converge in is identical.
        filtered = suppress_findings(findings)
        if filtered["dropped"]:
            self.say("observe",
                f"{len(filtered['dropped'])} finding(s) suppressed (generated code or test-file style)")
        findings = sorted(filtered["kept"], key=cmp_to_key(by_risk))
        self.state["findings"] = findings
        baseline = {f["fingerprint"] for f in findings}

        for r in runs:
            self.say("observe", f"{len(r['findings'])} finding(s) ({r['durationMs']}ms)")
        self.say("observe", f"Graph: {len(checks)} observer agent(s) ran concurrently")

        if self.config.get("dryRun"):
            self._dry_survey(findings, source_root)
            self.state["status"] = "converged"
            self.tracer.flush()
            return self.state

        max_iterations = int(self.config.get("maxIterations", 12))
        for i in range(1, max_iterations + 1):
            self.state["iteration"] = i
            target = self._next_target()
            if target is None:
                self.say("diagnose", "Nothing actionable left. Converged.")
                self.state["status"] = "converged"
                self.say("settle", "Converged.")
                self.tracer.flush()
                return self.state

            self.say("diagnose", f"Targeting {target['check']}: {target['summary']}")
            self._process(target, checks, source_root, baseline)

        self.state["status"] = "exhausted"
        self.say("settle", f"Iteration budget ({max_iterations}) reached")
        self.tracer.flush()
        return self.state

    def _run_all(self, checks, source_root):
        with ThreadPoolExecutor(max_workers=max(len(checks), 1)) as ex:
            return list(ex.map(lambda c: run_check(c, source_root), checks))

    def _next_target(self):
        skipped = 0
        for f in self.state["findings"]:
            if f["fingerprint"] in self.quarantined_this_run or self.ledger.is_exhausted(f):
                skipped += 1
                continue
            if skipped:
                self.say("diagnose", f"{skipped} finding(s) already quarantined — left for a human")
            return f
        return None

    # ------------------------------------------------------------- process

    def _process(self, target, checks, source_root, baseline):
        # Rules first: free, instant, and proven on these classes.
        candidates = propose_patches(target, source_root)

        # The model is consulted only for what the rules could not reach.
        if not candidates and self.provider:
            self.say("verify", f"No rule covers {target.get('code') or target['check']} — asking {self.provider.name}")
            try:
                candidates = self.provider.propose(target, source_root)
                self.say("verify", f"{self.provider.name} proposed {len(candidates)} candidate(s)")
                self.tracer.generation(
                    "propose",
                    getattr(self.provider, "last_usage", None),
                    check=target["check"], candidates=len(candidates),
                )
            except Exception as err:
                self.say("verify", f"Model proposer failed: {err}")
                candidates = []

        # Blast radius is decided now, from what the file is — a repair that
        # touches a module other files import changes code the loop is not
        # looking at, and the verify gate cannot catch that.
        graph = build_import_graph(source_root)
        for p in candidates:
            scope, importers = scope_of(graph, p["file"])
            p["scope"] = scope
            p["blastRadius"] = importers

        viable = self.ledger.prioritise(target["fingerprint"], candidates)
        if not viable:
            self.say("verify", "No untried patch available for this finding — quarantining for a human")
            self.quarantined_this_run.add(target["fingerprint"])
            self.record(target, None, "unverifiable", [])
            return

        patch = viable[0]

        # Three critic agents, fresh context, in parallel. Only a provider
        # that can judge is consulted; abstaining is "keep", which is safe
        # because the deterministic verify gate still runs afterwards.
        critique = getattr(self.provider, "critique", None) if self.provider else None
        if critique:
            def ask(question):
                try:
                    return critique(patch, target, question)
                except Exception as err:
                    self.say("verify", f"critic failed: {err}")
                    return None
            with ThreadPoolExecutor(max_workers=len(CRITIC_QUESTIONS)) as ex:
                verdicts = list(ex.map(ask, CRITIC_QUESTIONS))
            drops = sum(1 for v in verdicts if v and v.get("verdict") == "drop")
            if drops:
                self.say("verify", f"{drops}/{len(CRITIC_QUESTIONS)} checks voted drop — "
                          f"{'proceeding' if drops < 2 else 'rejecting'}")
            if drops >= 2:  # majority of three
                self.say("verify", "Critic agents rejected the patch — recording and moving on")
                self.record(target, patch, "unverifiable", [])
                return

        # A shared file is out of scope for an automatic fix, however cleanly
        # it would clear the check. Editing it moves code the loop is not
        # looking at, which is a decision its owner makes — and the verify
        # gate cannot catch this, because the finding really does clear.
        if patch["scope"] == "shared" and not self.config.get("allowShared"):
            rel = os.path.relpath(patch["file"], source_root).replace("\\", "/")
            self.say("verify", f"ESCALATED — {patch['rationale']}")
            self.say("verify",
                f"Not applied: {rel} is imported by {patch['blastRadius']} module(s). "
                "Editing it reaches past the defect. Re-run with --allow-shared to override.")
            self.record(target, patch, "unverifiable", [])
            return
        if patch["scope"] == "shared":
            self.say("verify", f"Shared-file change permitted by --allow-shared ({patch['blastRadius']} importer(s))")

        self.say("verify", f"Patch: {patch['rationale']}")

        restore = apply_edits([patch] + (patch.get("also") or []), source_root)
        outcome = None
        collateral = []
        collateral_detail = []
        runs = None
        try:
            v = verify_patch(checks, source_root, baseline, target["fingerprint"])
            outcome = v["outcome"]
            collateral = [f["fingerprint"] for f in v["collateral"]]
            collateral_detail = [f"{f['check']}: {f['summary']}" for f in v["collateral"]]
            runs = v["runs"]
            self.tracer.span(
                "verify",
                outcome=v["outcome"],
                collateral=len(v["collateral"]),
                durationMs=sum(r["durationMs"] for r in v["runs"]),
            )
        except Exception as err:
            # A failed verification proves nothing, so it cannot count as
            # success. Treat it as a failure and put the file back.
            self.say("verify", f"Verification failed: {err}")
            outcome = "unverifiable"

        if outcome != "committed":
            restore()
            self.say("verify", f"{outcome} — reverted")
            for c in collateral_detail:
                self.say("verify", f"  caused: {c}")
        else:
            self.say("verify", "committed — finding cleared with no collateral")
            # One commit per verified fix, so each can be read, kept, or
            # dropped on its own merits rather than as an all-or-nothing blob.
            if self.config.get("git"):
                try:
                    sha = commit_file(
                        source_root,
                        patch["file"],
                        f"Fix {target['check']}: {target['summary'][:60]}",
                        f"{target['summary']}\n\n{patch['rationale']}\n\n"
                        "Verified by re-running the checks after the change: "
                        "the finding cleared and no new finding appeared.",
                    )
                    if sha:
                        self.say("verify", f"  {sha} committed to git")
                except Exception as err:
                    self.say("verify", f"  git commit failed (the fix is still applied): {err}")

        self.record(target, patch, outcome, collateral)

        # Refresh findings from the verify run so cleared findings disappear
        # and collateral appears — the next target is chosen from reality.
        # The same suppression + risk ordering as the initial observe pass.
        if runs is not None:
            fresh = [f for r in runs for f in r["findings"]]
            filtered = suppress_findings(fresh)
            self.state["findings"] = sorted(filtered["kept"], key=cmp_to_key(by_risk))

    # ------------------------------------------------------------- dry survey

    def _dry_survey(self, findings, source_root):
        """Dry-run survey: report every finding's mechanical answer, write
        nothing."""
        with_patch = 0
        escalated = 0
        none = 0
        graph = build_import_graph(source_root)
        for f in findings:
            candidates = propose_patches(f, source_root)
            if not candidates:
                none += 1
                self.say("diagnose", f"NO PATCH   {f['check']}: {f['summary']}")
                continue
            scope, _ = scope_of(graph, candidates[0]["file"])
            shared = scope == "shared" and not self.config.get("allowShared")
            if shared:
                escalated += 1
                self.say("diagnose", f"ESCALATE   {f['check']}: {f['summary']}")
                rel = os.path.relpath(candidates[0]["file"], source_root).replace("\\", "/")
                self.say("diagnose", f"           → {rel} — shared file, not applied automatically")
            else:
                with_patch += 1
                self.say("diagnose", f"WOULD PATCH {f['check']}: {f['summary']}")
                rel = os.path.relpath(candidates[0]["file"], source_root).replace("\\", "/")
                self.say("diagnose", f"           → {rel} — {candidates[0]['rationale']}")
        self.say("settle",
            f"Dry run: {with_patch} patchable, {escalated} escalated, {none} with no mechanical answer. Nothing written.")

    # ------------------------------------------------------------- record

    def record(self, finding, patch, outcome, collateral):
        attempt = {
            "fingerprint": finding["fingerprint"],
            "patch": patch if patch else {
                "id": "none", "file": "", "find": "", "replace": "",
                "rationale": f"no candidate patch for {finding['check']}: {finding['summary']}",
                "scope": "local",
            },
            "outcome": outcome,
            "at": _iso_now(),
            "collateral": collateral,
            "provider": bool(self.provider),
        }
        self.ledger.record(attempt)
        self.state["attempts"].append(attempt)
